### ISSUE TITLE
2.5.0: openhab-webui components no longer distributed with openhab{-core}

### ISSUE THREAD

**Issue by tgurr** [openhab/openhab-distro#1019]
Looks like the webui components were outsorced to https://github.com/openhab/openhab-webui with https://github.com/openhab/openhab-core/commit/fb0470c6d10d46bc05f5a893fb0a8649849222d7, so in the recent openhab 2.5.0 release they're not longer shipped with openhab-core. This causes a problem for shipping openhab in distributions as previously we could package a working openhab with openhab{-core} and openhab-addons. Since openhab 2.5.0 openhab tries to download them on demand e.g. when starting up openhab, resulting in error messages and a non functional (not reachable via :8080) openhab when running openhab behind a proxy or a corporate network where openhab isn't allowed to download anything from the web. Is there a additional .kar file available with the webui components bundled, like for openhab-addons that we could ship to be able to package a working out of the box openhab 2.5.0 installation? I couldn't find anything useful on https://dl.bintray.com/openhab/mvn/org/openhab/distro/

**Comment by kaikreuzer** [comment]
> Indeed, the UI components should be part of the addons.kar.

I've just checked [openhab-addons-2.5.0.kar](https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab-addons%2F2.5.0%2Fopenhab-addons-2.5.0.kar) and I noticed that all UIs actually ARE already part of it.

So what exactly are you missing...?

**Comment by tgurr** [comment]
My openhab-addons-2.5.0.kar has the md5sum: ec107e92fbcd8c8f6993c2f550e1e4d0, is that correct?

Without additionally installing the mentioned openhab-webui jars at:

```
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.restdocs/2.5.0/org.openhab.ui.restdocs-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.paper/2.5.0/org.openhab.ui.paper-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.iconset.classic/2.5.0/org.openhab.ui.iconset.classic-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.homebuilder/2.5.0/org.openhab.ui.homebuilder-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.habpanel/2.5.0/org.openhab.ui.habpanel-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.habot/2.5.0/org.openhab.ui.habot-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.habmin/2.5.0/org.openhab.ui.habmin-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.dashboard/2.5.0/org.openhab.ui.dashboard-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.cometvisu/2.5.0/org.openhab.ui.cometvisu-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.cometvisu.php/2.5.0/org.openhab.ui.cometvisu.php-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.classic/2.5.0/org.openhab.ui.classic-2.5.0.jar
/usr/share/openhab2/runtime/system/org/openhab/ui/bundles/org.openhab.ui.basic/2.5.0/org.openhab.ui.basic-2.5.0.jar
```
I get the following error (openhab.log) when starting up openhab 2.5.0 for the first time (clean install):

```
2019-12-20 10:17:17.183 [ERROR] [ternal.service.BootFeaturesInstaller] - Error installing boot features
org.apache.karaf.features.internal.util.MultiException: Error:
	Error downloading mvn:org.openhab.ui.bundles/org.openhab.ui.dashboard/2.5.0
	at org.apache.karaf.features.internal.download.impl.MavenDownloadManager$MavenDownloader.<init>(MavenDownloadManager.java:91) ~[?:?]
	at org.apache.karaf.features.internal.download.impl.MavenDownloadManager.createDownloader(MavenDownloadManager.java:72) ~[?:?]
	at org.apache.karaf.features.internal.region.Subsystem.downloadBundles(Subsystem.java:457) ~[?:?]
	at org.apache.karaf.features.internal.region.Subsystem.downloadBundles(Subsystem.java:452) ~[?:?]
	at org.apache.karaf.features.internal.region.SubsystemResolver.resolve(SubsystemResolver.java:224) ~[?:?]
	at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:393) ~[?:?]
	at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1062) ~[?:?]
	at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:998) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_232]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_232]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_232]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_232]
	Suppressed: java.io.IOException: Error downloading mvn:org.openhab.ui.bundles/org.openhab.ui.dashboard/2.5.0
		at org.apache.karaf.features.internal.download.impl.AbstractRetryableDownloadTask.run(AbstractRetryableDownloadTask.java:77) ~[?:?]
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_232]
		at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_232]
		at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_232]
		at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_232]
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_232]
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_232]
		at java.lang.Thread.run(Thread.java:748) [?:1.8.0_232]
	Caused by: java.io.IOException: Error resolving artifact org.openhab.ui.bundles:org.openhab.ui.dashboard:jar:2.5.0: [Could not transfer artifact org.openhab.ui.bundles:org.openhab.ui.dashboard:jar:2.5.0 from/to openhab (https://dl.bintray.com/openhab/mvn/online-repo/2.5/): Connect to dl.bintray.com:443 [dl.bintray.com/52.29.129.139, dl.bintray.com/3.124.133.231] failed: connect timed out]
		at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.configureIOException(AetherBasedResolver.java:803) ~[?:?]
		at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:774) ~[?:?]
		at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:657) ~[?:?]
		at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:598) ~[?:?]
		at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:565) ~[?:?]
		at org.apache.karaf.features.internal.download.impl.MavenDownloadTask.download(MavenDownloadTask.java:52) ~[?:?]
		at org.apache.karaf.features.internal.download.impl.AbstractRetryableDownloadTask.run(AbstractRetryableDownloadTask.java:60) ~[?:?]
		... 7 more
		Suppressed: shaded.org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact org.openhab.ui.bundles:org.openhab.ui.dashboard:jar:2.5.0 from/to openhab (https://dl.bintray.com/openhab/mvn/online-repo/2.5/): Connect to dl.bintray.com:443 [dl.bintray.com/52.29.129.139, dl.bintray.com/3.124.133.231] failed: connect timed out
			at shaded.org.eclipse.aether.connector.basic.ArtifactTransportListener.transferFailed(ArtifactTransportListener.java:52) ~[?:?]
			at shaded.org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:368) ~[?:?]
			at shaded.org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:75) ~[?:?]
			at shaded.org.eclipse.aether.connector.basic.BasicRepositoryConnector$DirectExecutor.execute(BasicRepositoryConnector.java:642) ~[?:?]
			at shaded.org.eclipse.aether.connector.basic.BasicRepositoryConnector.get(BasicRepositoryConnector.java:262) ~[?:?]
			at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.performDownloads(DefaultArtifactResolver.java:489) ~[?:?]
			at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:390) ~[?:?]
			at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:215) ~[?:?]
			at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:192) ~[?:?]
			at shaded.org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifact(DefaultRepositorySystem.java:247) ~[?:?]
			at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:767) ~[?:?]
			at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:657) ~[?:?]
			at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:598) ~[?:?]
			at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:565) ~[?:?]
			at org.apache.karaf.features.internal.download.impl.MavenDownloadTask.download(MavenDownloadTask.java:52) ~[?:?]
			at org.apache.karaf.features.internal.download.impl.AbstractRetryableDownloadTask.run(AbstractRetryableDownloadTask.java:60) ~[?:?]
			at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_232]
			at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_232]
			at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_232]
			at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_232]
			at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_232]
			at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_232]
			at java.lang.Thread.run(Thread.java:748) [?:1.8.0_232]
		Caused by: shaded.org.apache.maven.wagon.TransferFailedException: Connect to dl.bintray.com:443 [dl.bintray.com/52.29.129.139, dl.bintray.com/3.124.133.231] failed: connect timed out
			at shaded.org.apache.maven.wagon.shared.http.AbstractHttpClientWagon.fillInputData(AbstractHttpClientWagon.java:1184) ~[?:?]
			at shaded.org.apache.maven.wagon.shared.http.AbstractHttpClientWagon.fillInputData(AbstractHttpClientWagon.java:1072) ~[?:?]
			at shaded.org.apache.maven.wagon.StreamWagon.getInputStream(StreamWagon.java:126) ~[?:?]
			at shaded.org.apache.maven.wagon.StreamWagon.getIfNewer(StreamWagon.java:88) ~[?:?]
			at shaded.org.apache.maven.wagon.StreamWagon.get(StreamWagon.java:61) ~[?:?]
			at shaded.org.eclipse.aether.transport.wagon.WagonTransporter$GetTaskRunner.run(WagonTransporter.java:567) ~[?:?]
			at shaded.org.eclipse.aether.transport.wagon.WagonTransporter.execute(WagonTransporter.java:435) ~[?:?]
			at shaded.org.eclipse.aether.transport.wagon.WagonTransporter.get(WagonTransporter.java:412) ~[?:?]
			at shaded.org.eclipse.aether.connector.basic.BasicRepositoryConnector$GetTaskRunner.runTask(BasicRepositoryConnector.java:456) ~[?:?]
			at shaded.org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:363) ~[?:?]
			... 21 more
		Caused by: shaded.org.apache.http.conn.ConnectTimeoutException: Connect to dl.bintray.com:443 [dl.bintray.com/52.29.129.139, dl.bintray.com/3.124.133.231] failed: connect timed out
			at shaded.org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:151) ~[?:?]
			at shaded.org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:373) ~[?:?]
			at shaded.org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:394) ~[?:?]
			at shaded.org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:237) ~[?:?]
			at shaded.org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:185) ~[?:?]
			at shaded.org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89) ~[?:?]
			at shaded.org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110) ~[?:?]
			at shaded.org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185) ~[?:?]
			at shaded.org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83) ~[?:?]
			at org.ops4j.pax.url.mvn.internal.wagon.ConfigurableHttpWagon.execute(ConfigurableHttpWagon.java:162) ~[?:?]
			at shaded.org.apache.maven.wagon.shared.http.AbstractHttpClientWagon.fillInputData(AbstractHttpClientWagon.java:1095) ~[?:?]
			at shaded.org.apache.maven.wagon.shared.http.AbstractHttpClientWagon.fillInputData(AbstractHttpClientWagon.java:1072) ~[?:?]
			at shaded.org.apache.maven.wagon.StreamWagon.getInputStream(StreamWagon.java:126) ~[?:?]
			at shaded.org.apache.maven.wagon.StreamWagon.getIfNewer(StreamWagon.java:88) ~[?:?]
			at shaded.org.apache.maven.wagon.StreamWagon.get(StreamWagon.java:61) ~[?:?]
			at shaded.org.eclipse.aether.transport.wagon.WagonTransporter$GetTaskRunner.run(WagonTransporter.java:567) ~[?:?]
			at shaded.org.eclipse.aether.transport.wagon.WagonTransporter.execute(WagonTransporter.java:435) ~[?:?]
			at shaded.org.eclipse.aether.transport.wagon.WagonTransporter.get(WagonTransporter.java:412) ~[?:?]
			at shaded.org.eclipse.aether.connector.basic.BasicRepositoryConnector$GetTaskRunner.runTask(BasicRepositoryConnector.java:456) ~[?:?]
			at shaded.org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:363) ~[?:?]
			... 21 more
		Caused by: java.net.SocketTimeoutException: connect timed out
			at java.net.PlainSocketImpl.socketConnect(Native Method) ~[?:1.8.0_232]
			at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[?:1.8.0_232]
			at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[?:1.8.0_232]
			at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[?:1.8.0_232]
			at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[?:1.8.0_232]
			at java.net.Socket.connect(Socket.java:607) ~[?:1.8.0_232]
			at shaded.org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:339) ~[?:?]
			at shaded.org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:142) ~[?:?]
			at shaded.org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:373) ~[?:?]
			at shaded.org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:394) ~[?:?]
			at shaded.org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:237) ~[?:?]
			at shaded.org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:185) ~[?:?]
			at shaded.org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89) ~[?:?]
			at shaded.org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110) ~[?:?]
			at shaded.org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185) ~[?:?]
			at shaded.org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83) ~[?:?]
			at org.ops4j.pax.url.mvn.internal.wagon.ConfigurableHttpWagon.execute(ConfigurableHttpWagon.java:162) ~[?:?]
			at shaded.org.apache.maven.wagon.shared.http.AbstractHttpClientWagon.fillInputData(AbstractHttpClientWagon.java:1095) ~[?:?]
			at shaded.org.apache.maven.wagon.shared.http.AbstractHttpClientWagon.fillInputData(AbstractHttpClientWagon.java:1072) ~[?:?]
			at shaded.org.apache.maven.wagon.StreamWagon.getInputStream(StreamWagon.java:126) ~[?:?]
			at shaded.org.apache.maven.wagon.StreamWagon.getIfNewer(StreamWagon.java:88) ~[?:?]
			at shaded.org.apache.maven.wagon.StreamWagon.get(StreamWagon.java:61) ~[?:?]
			at shaded.org.eclipse.aether.transport.wagon.WagonTransporter$GetTaskRunner.run(WagonTransporter.java:567) ~[?:?]
			at shaded.org.eclipse.aether.transport.wagon.WagonTransporter.execute(WagonTransporter.java:435) ~[?:?]
			at shaded.org.eclipse.aether.transport.wagon.WagonTransporter.get(WagonTransporter.java:412) ~[?:?]
			at shaded.org.eclipse.aether.connector.basic.BasicRepositoryConnector$GetTaskRunner.runTask(BasicRepositoryConnector.java:456) ~[?:?]
			at shaded.org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:363) ~[?:?]
			... 21 more
	Caused by: shaded.org.eclipse.aether.resolution.ArtifactResolutionException: Error resolving artifact org.openhab.ui.bundles:org.openhab.ui.dashboard:jar:2.5.0
		at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:413) ~[?:?]
		at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:215) ~[?:?]
		at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:192) ~[?:?]
		at shaded.org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifact(DefaultRepositorySystem.java:247) ~[?:?]
		at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:767) ~[?:?]
		at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:657) ~[?:?]
		at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:598) ~[?:?]
		at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:565) ~[?:?]
		at org.apache.karaf.features.internal.download.impl.MavenDownloadTask.download(MavenDownloadTask.java:52) ~[?:?]
		at org.apache.karaf.features.internal.download.impl.AbstractRetryableDownloadTask.run(AbstractRetryableDownloadTask.java:60) ~[?:?]
		... 7 more
```
With the jars installed manually everything it starting up and working fine on a clean install.

**Comment by kaikreuzer** [comment]
Ok, I'll try to analyse this in depth soon.
It seems this is a duplicate of https://github.com/openhab/openhab-distro/issues/961, right?

**Comment by tgurr** [comment]
@kaikreuzer yes, looks like that's the case. Thank you in advance for looking into this soon.

**Comment by kaikreuzer** [comment]
I think, I have found the culprit: It is this change: https://github.com/openhab/openhab-distro/pull/883

This causes Karaf to install the Dashboard even before the KAR is extracted and its content made available.

If you remove that line from `userdata/etc/org.apache.karaf.features.cfg`, the error goes away. As a result, you won't see the Dashboard in a fresh installation, though.

You can either install it on the openHAB console with
```
feature:install openhab-ui-dashboard
```
or, if you are anyhow using textual configuration, add `dashboard` to the `ui=` section in `conf/services/addons.cfg`:

```
# A comma-separated list of UIs to install (e.g. "ui = basic,paper")
ui = dashboard,paper,basic
```

Unfortunately, a fix cannot be included in the 2.5.1 addons, since the issue is within the distro, for which we do not plan to do a 2.5.1 release.

But I hope the above information is a suitable workaround for anybody wanting to do an offline installation.

Issue will be fixed for openHAB 3.
